bridge/tracker/tracker.go

```diff
@@ -0,0 +1,390 @@
+	//	stop subscription
+	t.cancelSubscription()
+
+	// 	stop processing headers
+	t.cancelEventTracking()
+
+	//	close rootchain client
+	if err := t.client.close(); err != nil {
+		t.logger.Error("cannot close rootchain client", "err", err)
+
+		return err
+	}
+
+	return nil
+}
+
+//	getEventChannel returns the tracker's event channel.
+func (t *Tracker) getEventChannel() <-chan []byte {
+	return t.eventCh
+}
+
+//	startSubscription handles the subscription object (provided by the rootchain client).
+func (t *Tracker) startSubscription(ctx context.Context, sub subscription) {
+	for {
+		select {
+		case err := <-sub.err():
+			t.logger.Error("subscription error", err)
+
+			t.cancelSubscription()
+			t.logger.Debug("cancelling subscription")
+
+			return
+		case <-ctx.Done():
+			if err := sub.unsubscribe(); err != nil {
+				t.logger.Error("cannot unsubscribe", "err", err)
+
+				return
+			}
+
+			t.logger.Debug("subscription stopped")
+
+			return
+		}
+	}
+}
+
+//	startEventTracking tracks each header (received by the subscription) for events.
+func (t *Tracker) startEventTracking(ctx context.Context) {
+	for {
+		select {
+		case header := <-t.headerCh:
+			t.trackHeader(header)
+		case <-ctx.Done():
+			t.logger.Debug("stopping header process")
+
+			return
+		}
+	}
+}
+
+//	loadDB creates a new database (or loads existing)
+//	for storing the last processed block's number by the tracker.
+func (t *Tracker) loadDB() (*leveldb.DB, error) {
+	//	get path
+	cwd, err := os.Getwd()
+	if err != nil {
+		return nil, err
+	}
+
+	//	create or load db
+	db, err := leveldb.OpenFile(cwd+"/event_tracker/last_block_number", nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return db, nil
+}
+
+//	trackHeader determines the range of block to query
+//	based on the blockNumber of the header and issues an appropriate
+//	eth_getLogs call. If an event of interest was emitted,
+//	it is sent to eventCh.
+func (t *Tracker) trackHeader(header *ethHeader) {
+	//	determine range of blocks to query
+	fromBlock, toBlock, ok := t.calculateRange(header)
+	if !ok {
+		//	we are returning here because the calculated
+		//	range was already queried or the chain is not
+		//	at the desired depth.
+		return
+	}
+
+	t.logger.Info("querying events within block range", "from", fromBlock, "to", toBlock)
+
+	//	fetch logs
+	logs := t.queryEvents(fromBlock, toBlock)
+	t.logger.Info("matched events", "num", len(logs))
+
+	//	notify each matched log event
+	t.notify(logs...)
+}
+
+//	calculateRange determines the next range of blocks
+//	to query for events.
+func (t *Tracker) calculateRange(header *ethHeader) (from, to uint64, ok bool) {
+	//	extract block number from header field
+	//	TODO: polygon-edge header
+	latestHeight := big.NewInt(0).SetUint64(header.Number)
+
+	//	or: TODO ethereum header
+	//latestHeight, _ := types.ParseUint256orHex(&header.Number)
+
+	//	check if block number is at required depth
+	confirmations := big.NewInt(0).SetUint64(t.confirmations)
```

Comment:
Why not keep confirmations as `*big.Int`? You only initialize it once anyways, but do this conversion every time for a new header

Reply:
[Changed](https://github.com/0xPolygon/polygon-edge/pull/386/commits/dd26ae7d3aaafcd50b168d1cab05b94552718c36)